Return a pair [type, what] from getExceptionMessage

hoodmane opened the pull request:
As discussed in #17362, this changes `getExceptionMessage` to return a pair, the exception type and the exception `what()` if available. If no exception description is available, the second entry will be undefined. 

This is an API break so we may want to give the new function a different name and keep the old one or something. The behavior of `__get_exception_terminate_message` should be unchanged. This change adds a few extra heap allocations but this is not supposed to be hot code.

aheejin reviewed (commented):
Thank you. Sorry for the delayed review. I don't usually check all the PRs so please feel free to cc me if necessary!

hoodmane commented:
Thanks for the review @aheejin

> Sorry for the delayed review. I don't usually check all the PRs so please feel free to cc me if necessary!

Right, I'd meant to @ you but apparently I didn't.

I think I addressed your comments.

aheejin commented on system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:
```diff
@@ -81,38 +88,49 @@ char* __get_exception_message(void* thrown_object, bool terminate=false) {
+
+// Returns a message saying that execution was terminated due to an exception.
+// This message is freshly malloc'd and should be freed.
+char* __get_exception_terminate_message(void *thrown_object) {
```
```suggestion
char* __get_exception_terminate_message(void* thrown_object) {
```
clang-format

aheejin commented on system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:
```diff
@@ -81,38 +88,49 @@ char* __get_exception_message(void* thrown_object, bool terminate=false) {
-             (terminate ? "terminating with uncaught exception of type %s: %s"
-                        : "exception of type %s: %s"),
-             type_name,
-             what);
+            "terminating with uncaught exception %s: %s",
+              type,
+              message);
```
```suggestion
    asprintf(
      &result, "terminating with uncaught exception %s: %s", type, message);
```
clang-format

aheejin commented on system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:
```diff
@@ -81,38 +88,49 @@ char* __get_exception_message(void* thrown_object, bool terminate=false) {
-  }
-
-  if (demangled_buf) {
-    free(demangled_buf);
+            "terminating with uncaught exception of type %s",
+            type);
```
```suggestion
    asprintf(&result, "terminating with uncaught exception of type %s", type);
```
clang-format

aheejin commented on system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:
```diff
@@ -81,38 +88,49 @@ char* __get_exception_message(void* thrown_object, bool terminate=false) {
+  free(type);
   return result;
 }
 
```
```suggestion
```
clang-format

aheejin commented on system/lib/libcxxabi/src/cxa_exception_emscripten.cpp:
```diff
@@ -71,7 +72,13 @@ void* __thrown_object_from_unwind_exception(
 }
 
-char* __get_exception_message(void* thrown_object, bool terminate=false) {
+// Puts information about thrown_object into type and message
+// `type` is the string representation of the type of the exception
+// `message` is the `exception->what()` method if one is found,
+// otherwise `message` will be `NULL`
+// `type` is malloc'd and should be freed after use.
+// `message` is either `NULL` or a malloc'd string which should be freed.
```
```suggestion
// Given a thrown_object, puts the information about its type and message into  
// 'type' and 'message' output parameters. 'type' will contain the string        
// representation of the type of the exception, e.g., 'int'. 'message' will      
// contain the result of 'std::exception::what()' method if the type of the      
// exception is a subclass of std::exception; otherwise it will be NULL. The     
// caller is responsible for freeing 'type' buffer and also 'message' buffer, if 
// it is not NULL.
```
I tried to add some more info; please let me know if this doesn't sound better 😅

hoodmane replied:
Yup it's an improvement thanks.

aheejin commented:
Hmm, I tried to add suggestions for the clang-format, but they included deleted lines, which is not supported. What I tried to say was this diff:
```diff
diff --git a/system/lib/libcxxabi/src/cxa_exception_emscripten.cpp b/system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
index 39ac0dc1e..cd1932a15 100644
--- a/system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
+++ b/system/lib/libcxxabi/src/cxa_exception_emscripten.cpp
@@ -111,26 +111,21 @@ void __get_exception_message(void* thrown_object, char** type, char** message) {
 
 // Returns a message saying that execution was terminated due to an exception.
 // This message is freshly malloc'd and should be freed.
-char* __get_exception_terminate_message(void *thrown_object) {
+char* __get_exception_terminate_message(void* thrown_object) {
   char* type;
   char* message;
   __get_exception_message(thrown_object, &type, &message);
   char* result;
   if (message != NULL) {
-    asprintf(&result,
-            "terminating with uncaught exception %s: %s",
-              type,
-              message);
+    asprintf(
+      &result, "terminating with uncaught exception %s: %s", type, message);
     free(message);
   } else {
-    asprintf(&result,
-            "terminating with uncaught exception of type %s",
-            type);
+    asprintf(&result, "terminating with uncaught exception of type %s", type);
   }
   free(type);
   return result;
 }
-
 }
 
 #endif // __USING_EMSCRIPTEN_EXCEPTIONS__ || __USING_WASM_EXCEPTIONS__
```

hoodmane commented:
I applied the clang-format changes.

aheejin commented:
Thanks! By the way I'm not sure why a bunch of CIs are failing. Can you try merging the latest main branch into this?

hoodmane commented:
Problem is when building with wasm:
```
error: undefined symbol: $getExceptionMessageCommon (referenced by $getExceptionMessage__deps: ['$getCppExceptionThrownObjectFromWebAssemblyException','$getExceptionMessageCommon'], referenced by top-level compiled C/C++ code)
```

aheejin reviewed (approved):
Thank you!
